### PR TITLE
Make Beacon an optional feature

### PR DESCRIPTION
**Pull request by peterbarker**
~14kB on SkyViper-v2450

### CODE REVIEW

File: libraries/AP_Beacon/AP_Beacon_Backend.h [hunk 1]
```diff
@@ -62,3 +69,5 @@ class AP_Beacon_Backend
 
     AP_HAL::UARTDriver *uart;
 };
+
+#endif
```
Comment (tridge) [inline libraries/AP_Beacon/AP_Beacon_Backend.h:73]: // AP_BEACON_ENABLED

File: libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp [hunk 2]
```diff
@@ -198,8 +198,12 @@ void NavEKF2_core::setAidingMode()
         // Check if airspeed data is being used
         bool airSpdUsed = (imuSampleTime_ms - lastTasPassTime_ms <= minTestTime_ms);
 
+#if AP_BEACON_ENABLED
         // Check if range beacon data is being used
         bool rngBcnUsed = (imuSampleTime_ms - lastRngBcnPassTime_ms <= minTestTime_ms);
+#else
+        const bool rngBcnUsed = false;
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp:205]: both const?

File: libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp [hunk 3]
```diff
@@ -1008,15 +1010,21 @@ void NavEKF2_core::selectHeightForFusion()
         activeHgtSource = HGT_SOURCE_BARO;
     } else if ((frontend->_altSource == 2) && ((imuSampleTime_ms - lastTimeGpsReceived_ms) < 500) && validOrigin && gpsAccuracyGood) {
         activeHgtSource = HGT_SOURCE_GPS;
+#if AP_BEACON_ENABLED
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:1013]: make it use HGT_SOURCE_BCN instead of 3?

Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:1013]: test with hgt source=3 without beacon compiled in, make sure it doesn't blowup

File: libraries/AP_NavEKF2/AP_NavEKF2_core.cpp [hunk 4]
```diff
@@ -290,6 +293,7 @@ void NavEKF2_core::InitialiseVariables()
     rngBcnFuseDataReportIndex = 0;
     memset(&rngBcnFusionReport, 0, sizeof(rngBcnFusionReport));
     last_gps_idx = 0;
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_core.cpp:295]: why last_gps_idx here?

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 5]
```diff
@@ -979,6 +980,7 @@ void NavEKF3_core::readRngBcnData()
     }
 
 }
+#endif
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:983]: endif marker

**Review by tridge — CHANGES_REQUESTED** [review]
discussed on call:
 - decided to make EKF2 only compile with beacon
 - will look at moving all beacon vars in ekf3 into bcn structure

File: libraries/AP_Beacon/AP_Beacon_Backend.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Beacon/AP_Beacon_Backend.h]: Done

File: libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp]: Done (in EK3)

File: libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp]: No longer playing around in EK2.

**Comment by peterbarker** [comment]
```
Board              AP_Periph  blimp  copter  heli  plane  rover  sub
Durandal                      0      0       0     0      0      0
Hitec-Airspeed     0                                             
KakuteH7-bdshot               0      0       0     0      0      0
MatekF405                     0      0       0     0      0      0
Pixhawk1-1M                   0      0       0     0      0      0
f103-QiotekPeriph  0                                             
f303-Universal     0                                             
revo-mini                     0      0       0     0      0      0
skyviper-v2450                       -13872                      
```

**Comment by peterbarker** [comment]
I've built SITL without beacon enabled and while several tests fail they're all understandable.  Running the same suite with ekf2 is also OK - several failures which are related to running under EK2, and the expected failures in beacon tests.  Nothing falling out of the sky

**Comment by peterbarker** [comment]
>     * decided to make EKF2 only compile with beacon

That's actually very difficult.  We cut EKF2 out through elision rather than having #ifdef s floating around in the AP_NavEKF2 source code.  That means that it wants to find `AP::dal().beacon()` - but we can't get that unless the actual beacon code is in place.

>     * will look at moving all beacon vars in ekf3 into bcn structure

I believe we though that would be good as a future PR.  I think that could be a great cleanup, but would complicate this one.

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 6]
```diff
@@ -901,6 +902,7 @@ void NavEKF2_core::readRngBcnData()
     rngBcnDataToFuse = storedRangeBeacon.recall(rngBcnDataDelayed,imuDataDelayed.time_ms);
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:902]: could we make this const bool when beacon not enabled?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp (hunk 5 above)
Comment (peterbarker) [reply libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp]: Done

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp:902]: Done

**Pull request merged**